### ISSUE TITLE
Ammonite support proposal

### ISSUE THREAD

**Issue by alexarchambault** [scalameta/metals-feature-requests#80]
**Is your feature request related to a problem? Please describe.**

This describes a possible implementation of Ammonite support for metals, in order to gather comments / opinions.

Adding this feature would involve changes in both Ammonite and metals, so it's better to ensure everyone is onboard on both the Ammonite and metals side before proceeding. If everyone agrees with the idea here, I (@alexarchambault) could manage to work on it from the Scala Center.

**Describe the solution you'd like**

See goals below.

**Describe alternatives you've considered**

Worksheets are a possible alternative, yet don't cover the same use cases as Ammonite scripts. We propose here to have metals handle Ammonite `.sc` scripts, offering the same kind of experience as e.g. sbt projects.

**Additional context**

This is related to https://github.com/scalameta/metals-feature-requests/issues/62. We propose to work on a way to address https://github.com/scalameta/metals-feature-requests/issues/62.

**Search terms**:

Ammonite, scripts, worksheets

**Comment by alexarchambault** [comment]
## Goals

### Support for code navigation for Ammonite scripts on par with sbt projects support

That would bring a number of benefits to both beginners and advanced users, making it easier to move code back-and-forth between Ammonite scripts and standard projects right from the IDE, and providing the IDE experience from scripts. See also https://contributors.scala-lang.org/t/proposal-simplifying-the-scala-getting-started-experience/2978.

A typical workfrom this should enable would look like
- open a `*.sc` Ammonite script in vscode,
- be proposed to import it in metals, and once it's imported,
- get at least
  - accurate diagnostics,
  - types on hover,
  - navigate to definitions in the same script, in imported scripts (via `import $file`), and in the script dependencies (`import $ivy`),
  - parameters hints,
  - etc.

## Non-goals

### Worksheet / notebook-like inlining of results.

The goal of this proposal is *not* to provide a notebook-like or worksheet-like experience to Ammonite scripts, even though adding support for this could be a possibility in the future, in later developments. We're focusing on scripts for medium sized code bases, with e.g. values wrapped in functions most of the time rather than defined at the top-level.

### Running scripts from metals

It is *not* a goal of this proposal to offer to run scripts `@main` methods or [entrypoints](https://ammonite.io/#MultipleMainMethods) right from metals. This could be added in subsequent developments though. For now, we expect users to keep running scripts from the command-line.

**Comment by alexarchambault** [comment]
## Proposed changes

### BSP support in Ammonite

We propose to have Ammonite
- depend on [bsp4j](https://github.com/build-server-protocol/build-server-protocol)
- have its [`Interpreter`](https://github.com/lihaoyi/Ammonite/blob/2aa3f84e9e8c214f765f9204ab5403d75f5e88d3/amm/interp/src/main/scala/ammonite/interp/Interpreter.scala)
  - accept a `enableBsp: Boolean` field
  - expose a `buildServerOpt: Option[BuildServer]`, which would be non-empty iff `enableBsp` is true,
- accept a `--bsp` argument, that would start Ammonite, enabling BSP communication either on its standard input / output or via a named pipe.

Note that having `Interpreter` expose a `BuildServer` would require some refactorings, to be able to compile scripts while not running / evaluating them.

## New project: ammonite-runner

In order to allow metals to start any version of Ammonite, on a variety of scala versions, we propose to create a new project, ammonite-runner. That project would consist in a library, allowing to:
- detect the scala and Ammonite versions of a script,
- fetch the classpath of that version of Ammonite for that scala version via [coursier interface](https://github.com/coursier/interface),
- start an Ammonite process with BSP enabled on the script.

Note that this project wouldn't depend on Ammonite itself. It would only dynamically fetch its classpath.

That project would have only few dependencies, making it unlikely to conflict with those of metals.

Detecting a script Ammonite and scala versions could rely on a comment like
```scala
// Ammonite 2.0.4, scala 2.12.10
```
at the beginning of the script, that ammonite-runner would look for. If that version is not found, we could possibly warn users about it.

## Handling Ammonite processes from metals

We propose to have `MetalsLanguageServer` manage (itself or via an auxiliary class) BSP connections to Ammonite processes.

A point to pay attention to here is that several BSP connections may now fill `buildTargets` (the main build tool one, plus the ones of the Ammonite scripts).

These Ammonite processes and connections would be started / created by ammonite-runner (see above), that metals would depend on.

**Comment by alexarchambault** [comment]
## Possible issues

- up to one Ammonite running in the background per opened script (so up to one JVM per opened script)
- "importing" scripts in metals feels a bit odd, but is probably necessary (scripts may have dependencies, that need to be fetched, …)
- could https://github.com/scalameta/metals-feature-requests/issues/48#issuecomment-541138944 be a problem? The Ammonite BSP connections would only add targets to `buildTargets`, and these would cover different sources than the main project.

**Comment by alexarchambault** [comment]
cc @lihaoyi (Ammonite author), @Baccata (who did some groundwork around Ammonite and bloop / metals)

**Comment by ckipp01** [comment]
There is actually another feature request for this, but since this one is so detailed, I'll just link it here -- https://github.com/scalameta/metals-feature-requests/issues/62

**Comment by Baccata** [comment]
Thanks for pinging. I had started to dive into ammonites guts last year to assess whether bsp implementation was achievable, before life caught up and forced me to defer all OSS work. 

Here's a couple things I imagined were gonna be hard last time I looked : 

### Dynamic compiler settings 

Ammonite exposes an API that allows to imperatively mutate the interpreter (and the underlying compiler) in prevision of the next compile-unit. This means that adding coursier resolvers or changing compile options that would be reflected in editor via metals would either be impossible, or would require running upstream scripts. 

Any refactoring that would allow to compile without evaluating would necessarily change the semantics of ammonite, unless you accept a half-satisfying change that would avoid running the currently edited script but would still run its upstreams to aggregate the metadata necessary to the implementation of a bsp server. Otherwise you'll have to alienate a few metals users by implementing bsp support without catering to dynamic compiler settings. 

### The case of mill 

One of the usecases of ammonite is of course, the mill build tool . Mill differs from ammonite in the header code that is added to the script before it gets compiled (as well as the initial classpath the interpreter gets loaded with). To have a truly satisfying solution for ammonite support, you'd have to have your `ammonite-runner` component be able to distinguish between the ammonite case and the mill case, and have both `mill` and `ammonite` support a `--bsp` option. 

That might not be something you need to solve right this second, but you might want to have a quick look at how mill calls upon ammonite to avoid making it hard to address this usecase later on.

**Comment by alexarchambault** [comment]
Thanks for your comment @Baccata!

About the dynamic compiler settings, I'd like to push more for adding them via imports, like
```scala
import $scalac.`-Ysomething`
```
Same for adding repositories. Overall, if we don't run scripts, we can't really handle well `@` sections in them (because prior sections may do dynamic stuff like adding repositories, adding fancy dependencies via `interp.load.ivy`, change compiler settings, etc.) It'd be better to handle these via imports as much as possible, adding support for more import syntaxes if necessary.

About mill, I'm aware it does custom stuff prior to evaluating `build.sc`. I guess it's out of the scope of this in a first time, unless it turns out it's straightforward to handle that from ammonite-runner (I don't remember the details of what mill changes).

Also, mill usually has its own background process, maybe it'd be interesting to special case its scripts more (its bootstrap script could accept `--bsp` for example, and we would just run that).

**Comment by Baccata** [comment]
> It'd be better to handle these via imports as much as possible, adding support for more import syntaxes if necessary. 

Yup, totally agree there. However people are using dynamic compiler settings and the required refactoring will likely have to respect the current semantics and make them cohabit with no-exec ones. 

I don't think it'll be that useful, but here's a closed PR of some work I had started on what I thought (at the time) was gonna be required for bsp implementation : https://github.com/lihaoyi/Ammonite/pull/979
